### DISCUSSION TITLE
[feature-request]: evaluate should support apply()

### DISCUSSION THREAD

**Discussion by tmtron** [josdejong/mathjs#3041] [Design decisions]
**Describe the bug**
Currently (v11.6.0) [apply](https://mathjs.org/docs/reference/functions/apply.html) not work in evaluation mode. It would be good if that is supported.

see also [comment in #2908](https://github.com/josdejong/mathjs/issues/2908#:~:text=That%20is%20an%20unfortunate%20side%20effect) by @josdejong
>That is an unfortunate side effect: the property apply is considered an unsafe property because it exists as a prototype method of Function: https://github.com/josdejong/mathjs/blob/develop/src/utils/customs.js#L52-L84. If we want to see if we can improve on that, let's discuss in a separate topic.

**To Reproduce**
```
apply([[1, 2], [3, 4]], 1, sum)
Error: No access to property "apply"
```

**Comment by josdejong** [comment]
Anyone able to help think this though?

**Reply by dvd101x** [reply]
Hi, I found a few references

In numpy that function is called [apply_along_axis](https://numpy.org/doc/2.0/reference/generated/numpy.apply_along_axis.html#numpy.apply_along_axis) 

Also these functions can be extended to work on multiple axes or dimensions and there are references in Julia and numpy
- **Julia:** [mapslices](https://www.jlhub.com/julia/manual/en/function/mapslices)
- **numpy:** [apply_over_axes](https://numpy.org/doc/stable/reference/generated/numpy.apply_over_axes.html)

One possibility is rename `apply` to something like `applyOnDim`, `reduceDim`, `mapSlice`, etc. In a way to bring more meaning to the name of the function.

Out of scope but I think there is a possibility to use it for multiple dimensions like `applyOnDims` when the dim is not a number but an array of numbers.

**Reply by dvd101x** [reply]
Isn't `apply` very similar to `reduce` in **collection.js** ?

https://github.com/josdejong/mathjs/blob/6bcb26ca455a1d0f5a0439259c28aef256ed5911/src/utils/collection.js#L67-L88

Maybe there is an opportunity to join them.

**Reply by josdejong** [reply]
Thanks for looking into this David. 

It makes sense to me to rename the function, or create an alias for it to keep things backward compatible. I don't have strong feelings about the best name for the function, I like `mapSlice` and `applyOnDim` for example. Do you have a preference?

Good idea to see if we can unify the code of `apply` and `reduce`!

**Reply by dvd101x** [reply]
- `mapSlice` for me it has more meaning, except that the mapping function that is being applied to every slice of the array should reduce one dimension of each slice and I don't think that's implied by the name.
- `applyOnDim` I like that it maintains the reference to the current name apply, and the change form `apply` to `applyOnDim` sounds similar to what happen between `eval` and `evaluate`. Also is not implied in the name apply what is the intention of the callback (reduce one dimension). Also I'm not sure about about `dim` as in the documentation I see in some functions the full word `dimension` is used.
- `aggregate` is also a good fit and found a [nice explanation in pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.aggregate.html). The negative part is that it might be more used in database management.

**Reply by josdejong** [reply]
Shall we go for `mapSlice`?

**Reply by dvd101x** [reply]
Yes, I think it's a good name. Maybe it should be `mapSlices` regardles if it's a single dimension or multiple dimensions because even if it's a single dimension you can still have many slices of the matrix.

**Reply by josdejong** [reply]
Sounds good to me :)

**Comment by gwhitney** [comment]
Since this is more of an open-ended architecture discussion than a specific presently-actionable issue, moving to Discussions.

**Comment by gwhitney** [comment]
Is there any reason we can't make an exception to the safety check for `apply`?

**Reply by josdejong** [reply]
I guess that would be possible. However in order to prevent shooting ourselves in the foot some day with it, I think I prefer renaming the function `apply`.

**Reply by gwhitney** [reply]
ok, gotcha. The julia name for this operation is mapslices, so I agree the new name should either be mapslices or mapSlices depending on which casing you like better, with perhaps .apply() left as an alias to avoid a breaking change (at least until the next major version bump).

**Reply by gwhitney** [reply]
Looking over the current naming of functions, `mapSlices()` is clearly more consistent with other mathjs names, so preparing a PR that renames `apply()` to `mapSlices()` but makes no effort to unify or share more code with `reduce()`; we can open a new discussion for that point when this is resolved. I will leave `apply()` as a deprecated synonym for `mapSlices()` to avoid a breaking change; it could be removed in a future major-version bump if so desired.

**Reply by dvd101x** [reply]
I'm working on something related to unifying `map` and `forEach`, I can look into unifying `reduce` in case I can make it work.

**Comment by josdejong** [comment]
The function `apply` is now renamed to `mapSlices` in `v14.2.0`.
